### PR TITLE
Fix #10648: Fixes of syntax documentation

### LINKED ISSUE #10648: Documentation is unclear about new brace-less syntax
The documentation for the new syntax - arguably, a very important change in Scala 3 - is confusing.

The file is https://github.com/lampepfl/dotty/blob/master/docs/docs/reference/other-new-features/indentation.md and the relevant section is titled "Optional braces around template bodies". https://dotty.epfl.ch/docs/reference/other-new-features/indentation.html#optional-braces-around-template-bodies

First, it explains that the change introduces a colon at end of line. Then it gives examples, some of which do not involve any colon at all:

```
With these new rules, the following constructs are all valid:

...

type T = A:
  def f: Int

given [T](using Ord[T]): Ord[List[T]] with
  def compare(x: List[T], y: List[T]) = ???

extension (xs: List[Int])
  def second: Int = xs.tail.head

...

```

In the example with `type T = A:` it is not clear what is being done; is this `type T = A with Object { def f: Int }` or something else?

But the next two examples are `given ... with` and `extension`, which do not involve a colon at the end of line. The first one is the `given ... with`, and it is confusing since the `with` keyword is never explained. Actually, this use of the `with` keyword seems to be never mentioned in the documentation.

The example with `extension` does not have a colon at the end of line. Is this an error?

Please make the documentation complete in these crucial aspects. I would have pushed a PR directly for this documentation file, except that I don't actually know what would be the right explanation for those examples.

### PR DESCRIPTION
**Pull request by odersky**
Fix documentation about indentation syntax. Also fix syntax summary (globally).

The fixes omit documenting : in extension blocks. These are disabled
in a separate commit.

There are two reasons for omitted in : in extensions:

 - they are not needed, and there should be one way to do things
 - they are confusing since everywhere else a : at eol starts a
new scope, but here it doesn't.

### CODE REVIEW

**Comment by odersky** [comment]
There are now two syntax.md files, one in docs/reference, the other in docs/internal.

 - The file in docs/reference contains only the official 3.0 syntax and omits the column with tree node info.
 - The file in docs/internal contains also deprecated and experimental syntax and adds the tree node info column.

It will be a bit painful to keep the two in sync, but it's better to have amn official, stable syntacx summary in the official documentation.

File: docs/docs/reference/other-new-features/indentation.md [hunk 1]
```diff
@@ -145,14 +136,15 @@ package p:
...
 The syntax changes allowing this are as follows:
 
 ```
-TemplateBody ::=  [colonEol] ‘{’ [SelfType] TemplateStat {semi TemplateStat} ‘}’
-EnumBody     ::=  [colonEol] ‘{’ [SelfType] EnumStat {semi EnumStat} ‘}’
-Packaging    ::=  ‘package’ QualId [colonEol] ‘{’ TopStatSeq ‘}’
-RefinedType  ::=  AnnotType {[colonEol] Refinement}
+Template    ::=  InheritClauses [colonEol] [TemplateBody]
+EnumDef     ::=  id ClassConstr InheritClauses [colonEol] EnumBody
+Packaging   ::=  ‘package’ QualId [nl | colonEol] ‘{’ TopStatSeq ‘}’
```
Comment (b-studios) [inline docs/docs/reference/other-new-features/indentation.md:146]: Maybe I am missing something, but I cannot recognize the `colonEol` in the parser code for packages:

https://github.com/lampepfl/dotty/blob/44514a34943134f7072881c63502d3f773204f1b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala#L3720-L3724

Is `rewriteWithColon` doing this?

The comment above also does not reflect the colon: 

https://github.com/lampepfl/dotty/blob/44514a34943134f7072881c63502d3f773204f1b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala#L3718

File: tests/neg/i5773.scala [hunk 2]
```diff
@@ -23,7 +23,7 @@ object Main {
 21 |    println(1 appendS 2)
    |            ^^^^^^^^^
    |value appendS is not a member of Int.
-   |An extension method was tried, but could not be fully constructed:
```
Comment (b-studios) [inline tests/neg/i5773.scala:26]: Like the change in the source, is this intended?

**Review by b-studios — COMMENTED** [review]
Copying the two syntax documents indeed leads to additional maintenance effort. Should we just present the diff in internals and link to the official one?

File: compiler/src/dotty/tools/dotc/transform/ExtensionMethods.scala [hunk 3]
```diff
@@ -190,7 +190,7 @@ object ExtensionMethods {
         // See the documentation of `memberSignature` to understand why `.stripPoly.ensureMethodic` is needed here.
         candidates filter (c => FullParameterization.memberSignature(c.info) == imeth.info.stripPoly.ensureMethodic.signature)
       assert(matching.nonEmpty,
-       i"""no extension method found for:
+       i"""no extension method found for
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/transform/ExtensionMethods.scala:193]: Well spotted. Yes, that was my regex search misfiring...

File: docs/docs/reference/other-new-features/indentation.md (hunk 1 above)
Comment (odersky) [reply docs/docs/reference/other-new-features/indentation.md:146]: It's in `possibleTemplateStart()`. I added a test to verify that `:` is indeed significant.

**Comment by b-studios** [comment]
@odersky there are also still occurrences of `:` in the documentation on extension methods. I can fix those and push to this branch if you want.

**Comment by odersky** [comment]
@b-studios I think I got them all now. I was looking for "  extension.*:$" so extensions starting in the first column slipped through.

File: docs/docs/reference/contextual/extension-methods.md [hunk 4]
```diff
@@ -103,7 +103,7 @@ a single extension and enclose all methods in braces or an indented region follo
 Example:
```
Comment (b-studios) [inline docs/docs/reference/contextual/extension-methods.md:103]: The text above this line would also need to be adjusted.

Comment (odersky) [reply docs/docs/reference/contextual/extension-methods.md:103]: Well spotted. I'll update the doc.

**Comment by japgolly** [comment]
> they are confusing since everywhere else a : at eol starts a new scope, but here it doesn't.

I'm trying hard to understand this reasoning but I can't make sense of it. Pretty much everywhere else we have trailing `:`s, what's special about extension blocks? Even the comment above that it doesn't create a new scope makes no sense to me. But.... of course it creates a new scope. Everything indented under `extension` is defined within that `extension` scope. I'm sure the back and forth much be annoying but I urge you to please reconsider this. I can't explain emphatically enough how confusing this is.

**Pull request merged**